File: swift/03 Vertex and Index Buffers/finished/HelloTriangle/Renderer.swift
//
//  Renderer.swift
//  HelloTriangle
//
//  Created by Andrew Mengede on 27/2/2022.
//

import MetalKit

class Renderer: NSObject, MTKViewDelegate {
    
    var parent: ContentView
    var device: MTLDevice!
    var commandQueue: MTLCommandQueue!
    var pipeline: MTLRenderPipelineState
    var triangleMesh: MTLBuffer
    var quadMesh: Mesh
    
    init(_ parent: ContentView) {
        
        self.parent = parent
        if let device = MTLCreateSystemDefaultDevice() {
            self.device = device
        }
        self.commandQueue = device.makeCommandQueue()
        
        pipeline = build_pipeline(device: device)
        
        let meshBuilder = MeshFactory(device: device)
        triangleMesh = meshBuilder.make_triangle()
        quadMesh = meshBuilder.make_quad()
        
        super.init()
    }
    
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        
    }
    
    func draw(in view: MTKView) {
        
        guard let drawable = view.currentDrawable else {
            return
        }
        
        let commandBuffer = commandQueue.makeCommandBuffer()!
        
        let renderPassDescriptor = view.currentRenderPassDescriptor
        renderPassDescriptor?.colorAttachments[0].clearColor = MTLClearColorMake(0, 0.5, 0.5, 1.0)
        renderPassDescriptor?.colorAttachments[0].loadAction = .clear
        renderPassDescriptor?.colorAttachments[0].storeAction = .store
        
        let renderEncoder = commandBuffer
            .makeRenderCommandEncoder(descriptor: renderPassDescriptor!)!
        
        renderEncoder.setRenderPipelineState(pipeline)
        
        renderEncoder.setVertexBuffer(quadMesh.vertexBuffer, offset: 0, index: 0)
        renderEncoder.drawIndexedPrimitives(type: .triangle, indexCount: quadMesh.indexCount, 
                                            indexType: .uint16, indexBuffer: quadMesh.indexBuffer,
                                            indexBufferOffset: 0)
        
        renderEncoder.setVertexBuffer(triangleMesh, offset: 0, index: 0)
        renderEncoder.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 3)
        
        renderEncoder.endEncoding()
        
        commandBuffer.present(drawable)
        commandBuffer.commit()
    }
}

File: cpp/01 hello metal/finished/hello_metal_cpp/view/renderer.cpp
//
//  renderer.cpp
//  hello_metal_cpp
//
//  Created by Andrew Mengede on 15/9/2023.
//
#include "renderer.h"

Renderer::Renderer(MTL::Device* device):
device(device->retain()),
commandQueue(device->newCommandQueue())
{}

void Renderer::draw(MTK::View* view)
{
    const release_ptr<NS::AutoreleasePool> pool(NS::AutoreleasePool::alloc()->init());

    MTL::CommandBuffer* commandBuffer = commandQueue->commandBuffer();
    MTL::RenderPassDescriptor* renderPass = view->currentRenderPassDescriptor();
    MTL::RenderCommandEncoder* encoder = commandBuffer->renderCommandEncoder(renderPass);
    encoder->endEncoding();
    commandBuffer->presentDrawable(view->currentDrawable());
    commandBuffer->commit();
}
